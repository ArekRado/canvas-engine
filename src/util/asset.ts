import { State } from '../type'

type AddSprite = (param: { state: State; name: string; src: string }) => State
export const addSprite: AddSprite = ({ state, name, src }): State => ({
  ...state,
  asset: {
    sprite: [
      ...state.asset.sprite,
      {
        src,
        name,
      },
    ],
  },
})

type RemoveSprite = (param: { state: State; name: string }) => State
export const removeSprite: RemoveSprite = ({ state, name }): State => ({
  ...state,
  asset: {
    sprite: state.asset.sprite.filter((item) => item.name === name),
  },
})
