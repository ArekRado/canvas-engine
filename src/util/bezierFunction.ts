export type TimingFunction =
  | 'Linear'
  | 'EaseInQuad'
  | 'EaseOutQuad'
  | 'EaseInOutQuad'
  | 'EaseInCubic'
  | 'EaseOutCubic'
  | 'EaseInOutCubic'
  | 'EaseInQuart'
  | 'EaseOutQuart'
  | 'EaseInOutQuart'
  | 'EaseInQuint'
  | 'EaseOutQuint'
  | 'EaseInOutQuint'
  | 'CubicBezier' // todo  (number, number, number, number);

export const linear = (t) => t
// accelerating from zero velocity
export const easeInQuad = (t) => t * t
// decelerating to zero velocity
export const easeOutQuad = (t) => t * (2.0 - t)
// acceleration until halfway, then deceleration
export const easeInOutQuad = (t) =>
  t < 0.5 ? 2.0 * t * t : -1.0 + (4.0 - 2.0 * t) * t
// accelerating from zero velocity
export const easeInCubic = (t) => t * t * t
// decelerating to zero velocity
export const easeOutCubic = (t) => (t - 1.0) * t * t + 1.0
// acceleration until halfway, then deceleration
export const easeInOutCubic = (t) =>
  t < 0.5
    ? 4.0 * t * t * t
    : (t - 1.0) * (2.0 * t - 2.0) * (2.0 * t - 2.0) + 1.0
// accelerating from zero velocity
export const easeInQuart = (t) => t * t * t * t
// decelerating to zero velocity
export const easeOutQuart = (t) => 1.0 - (t - 1.0) * t * t * t
// acceleration until halfway, then deceleration
export const easeInOutQuart = (t) =>
  t < 0.5 ? 8.0 * t * t * t * t : 1.0 - 8.0 * (t - 1.0) * t * t * t
// accelerating from zero velocity
export const easeInQuint = (t) => t * t * t * t * t
// decelerating to zero velocity
export const easeOutQuint = (t) => 1.0 + (t - 1.0) * t * t * t * t
// acceleration until halfway, then deceleration
export const easeInOutQuint = (t) =>
  t < 0.5 ? 16.0 * t * t * t * t * t : 1.0 + 16.0 * (t - 1.0) * t * t * t * t

// export const Bezier =  p = (1-t)^3 *P0 + 3*t*(1-t)^2*P1 + 3*t^2*(1-t)*P2 + t^3*P3

export const getValue = (
  timingFunction: TimingFunction,
  value: number,
): number => {
  switch (timingFunction) {
    case 'Linear':
      return linear(value)
    case 'EaseInQuad':
      return easeInQuad(value)
    case 'EaseOutQuad':
      return easeOutQuad(value)
    case 'EaseInOutQuad':
      return easeInOutQuad(value)
    case 'EaseInCubic':
      return easeInCubic(value)
    case 'EaseOutCubic':
      return easeOutCubic(value)
    case 'EaseInOutCubic':
      return easeInOutCubic(value)
    case 'EaseInQuart':
      return easeInQuart(value)
    case 'EaseOutQuart':
      return easeOutQuart(value)
    case 'EaseInOutQuart':
      return easeInOutQuart(value)
    case 'EaseInQuint':
      return easeInQuint(value)
    case 'EaseOutQuint':
      return easeOutQuint(value)
    case 'EaseInOutQuint':
      return easeInOutQuint(value)
    case 'CubicBezier':
      return linear(value)
  }
}
