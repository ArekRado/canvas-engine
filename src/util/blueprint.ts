import { AssetBlueprint, Blueprint, State } from '../type'

const syncComponentBlueprint = (assetBlueprint: AssetBlueprint) => (
  state: State,
  blueprint: Blueprint,
): State => {
  const entityId = blueprint.entityId

  Object.entries(assetBlueprint.data).forEach(([componentKey, value]) => {
    state.component[componentKey][entityId] = {
      ...value,
      entityId: blueprint.entityId,
    }
  })

  return state
}

export const syncBlueprint = (state: State): State => {
  const blueprintComponent: Blueprint[] = Object.values(
    state.component.blueprint,
  )

  return state.asset.blueprint.reduce(
    (acc, assetBlueprint) =>
      blueprintComponent.reduce(syncComponentBlueprint(assetBlueprint), acc),
    state,
  )
}