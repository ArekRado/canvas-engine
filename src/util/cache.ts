export const createCacheContainer = <Data>() => {
  let cache: Record<string, Data> = {}

  return {
    reset: () => (cache = {}),
    set: (key: string, setter: () => Data) => {
      const oldValue = cache[key]

      if (oldValue) {
        return oldValue
      } else {
        const newValue = setter()
        cache[key] = newValue
        return newValue
      }
    },
  }
}
