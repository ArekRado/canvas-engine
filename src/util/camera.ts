import { camera as cameraDraw } from '../draw/camera'
import { Camera, State } from '../type'

type SetCamera = (params: { state: State; camera: Camera }) => State
export const setCamera: SetCamera = ({ state, camera }) => {
  state.regl && cameraDraw({ camera, regl: state.regl })

  return {
    ...state,
    camera,
  }
}
