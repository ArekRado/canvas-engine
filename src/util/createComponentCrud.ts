import { Entity } from '../type'
import { createStore } from './store'

export const createComponentCrud = <Component>({
  name,
  store,
}: {
  name: string
  store: ReturnType<typeof createStore>
}) => ({
  getComponent: (entity: Entity) => store.getComponent<Component>(name, entity),
  createComponent: (entity: Entity, data: Component) =>
    store.createComponent<Component>(name, entity, data),
  updateComponent: (
    entity: Entity,
    update: (component: Component) => Partial<Component>,
  ) => store.updateComponent<Component>(name, entity, update),

  removeComponent: (entity: Entity) => store.removeComponent(name, entity),
})
