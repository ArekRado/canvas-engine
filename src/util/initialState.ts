import { State } from '../type'
import { animationSystem } from '../system/animation'
import { collideBoxSystem } from '../system/collideBox'
import { drawSystem } from '../system/draw'
import { transformSystem } from '../system/transform'

const v1: State = {
  entity: [],
  component: {
    transform: {},
    sprite: {},
    animation: {},
    collideBox: {},
    collideCircle: {},
  },
  asset: {
    sprite: [],
  },
  time: {
    timeNow: 0.0,
    delta: 0.0,
  },
  mouse: {
    buttons: 0,
    position: [0, 0],
  },
  system: {},
  isDebugInitialized: false,
  isDrawEnabled: false,
}

const v2 = transformSystem(v1)
const v3 = drawSystem(v2)
const v4 = collideBoxSystem(v3)
const v5 = animationSystem(v4)

export const initialState = v5
