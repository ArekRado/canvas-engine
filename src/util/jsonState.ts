import { State } from '../type'

export const jsonToState = (jsonString: string, state: State): State => {
  const js = JSON.parse(jsonString)

  return {
    ...state,
    ...js,
  }
}

export const stateToJson = (state: State): string =>
  JSON.stringify({
    entity: state.entity,
    component: state.component,
    time: state.time,
    asset: state.asset,
    mouse: state.mouse,
    isDebugInitialized: state.isDebugInitialized,
    isDrawEnabled: state.isDrawEnabled,
  })
