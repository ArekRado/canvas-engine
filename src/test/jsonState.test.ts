import 'regenerator-runtime/runtime'
import { initialState } from '../util/initialState'
import { jsonToState, stateToJson } from '../util/jsonState'

describe('jsonState', () => {
  it('jsonToState ', () => {
    const jsonString = stateToJson(initialState)
    const state = jsonToState(jsonString, initialState)

    expect(state.entity).toMatchSnapshot()
    expect(state.component).toMatchSnapshot()
    expect(state.time).toMatchSnapshot()
    expect(state.asset).toMatchSnapshot()
    expect(state.mouse).toMatchSnapshot()
    expect(state.isDebugInitialized).toMatchSnapshot()
    expect(state.isDrawEnabled).toMatchSnapshot()
  })

  it('jsonToState', () => {
    const jsonString = stateToJson(initialState)

    expect(jsonString).toMatchSnapshot()
  })
})
