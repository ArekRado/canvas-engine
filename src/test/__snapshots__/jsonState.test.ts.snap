// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonState jsonToState  1`] = `Array []`;

exports[`jsonState jsonToState  2`] = `
Object {
  "animation": Object {},
  "collideBox": Object {},
  "collideCircle": Object {},
  "mouseInteraction": Object {},
  "sprite": Object {},
  "transform": Object {},
}
`;

exports[`jsonState jsonToState  3`] = `
Object {
  "delta": 0,
  "previousTimeNow": 0,
  "timeNow": 0,
}
`;

exports[`jsonState jsonToState  4`] = `
Object {
  "blueprint": Array [],
  "sprite": Array [],
}
`;

exports[`jsonState jsonToState  5`] = `
Object {
  "buttons": 0,
  "isButtonDown": false,
  "isButtonUp": false,
  "isMoving": false,
  "lastClick": Object {
    "buttons": 0,
    "timestamp": -1,
  },
  "position": Array [
    0,
    0,
  ],
}
`;

exports[`jsonState jsonToState  6`] = `false`;

exports[`jsonState jsonToState  7`] = `false`;

exports[`jsonState jsonToState 1`] = `"{\\"entity\\":[],\\"component\\":{\\"transform\\":{},\\"sprite\\":{},\\"animation\\":{},\\"collideBox\\":{},\\"collideCircle\\":{},\\"mouseInteraction\\":{}},\\"time\\":{\\"previousTimeNow\\":0,\\"timeNow\\":0,\\"delta\\":0},\\"asset\\":{\\"sprite\\":[],\\"blueprint\\":[]},\\"mouse\\":{\\"buttons\\":0,\\"position\\":[0,0],\\"isButtonUp\\":false,\\"isButtonDown\\":false,\\"isMoving\\":false,\\"lastClick\\":{\\"timestamp\\":-1,\\"buttons\\":0}},\\"isDebugInitialized\\":false,\\"isDrawEnabled\\":false}"`;
